File: app/src/main/java/org/mozilla/focus/tabs/tabtray/TabsSessionModel.java
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 4; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.focus.tabs.tabtray;

import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.view.View;
import android.webkit.GeolocationPermissions;
import android.webkit.ValueCallback;
import android.webkit.WebChromeClient;

import org.mozilla.focus.BuildConfig;
import org.mozilla.rocket.tabs.Session;
import org.mozilla.rocket.tabs.SessionManager;
import org.mozilla.rocket.tabs.TabView;
import org.mozilla.rocket.tabs.TabsChromeListener;
import org.mozilla.rocket.tabs.TabsViewListener;

import java.util.ArrayList;
import java.util.List;

class TabsSessionModel implements TabTrayContract.Model {
    @NonNull
    private SessionManager sessionManager;
    private OnTabModelChangedListener onTabModelChangedListener;

    private List<Session> tabs = new ArrayList<>();

    TabsSessionModel(@NonNull SessionManager sessionManager) {
        this.sessionManager = sessionManager;
    }

    @Override
    public void loadTabs(OnLoadCompleteListener listener) {
        tabs.clear();
        tabs.addAll(sessionManager.getTabs());

        if (listener != null) {
            listener.onLoadComplete();
        }
    }

    @Override
    public List<Session> getTabs() {
        return tabs;
    }

    @Override
    public Session getFocusedTab() {
        return sessionManager.getFocusSession();
    }

    @Override
    public void switchTab(int tabPosition) {
        if (tabPosition >= 0 && tabPosition < tabs.size()) {
            Session target = tabs.get(tabPosition);
            List<Session> latestTabs = sessionManager.getTabs();
            boolean exist = latestTabs.indexOf(target) != -1;
            if (exist) {
                sessionManager.switchToTab(target.getId());
            }
        } else {
            if (BuildConfig.DEBUG) {
                throw new ArrayIndexOutOfBoundsException("index: " + tabPosition + ", size: " + tabs.size());
            }
        }
    }

    @Override
    public void removeTab(int tabPosition) {
        if (tabPosition >= 0 && tabPosition < tabs.size()) {
            sessionManager.dropTab(tabs.get(tabPosition).getId());
        } else {
            if (BuildConfig.DEBUG) {
                throw new ArrayIndexOutOfBoundsException("index: " + tabPosition + ", size: " + tabs.size());
            }
        }
    }

    @Override
    public void clearTabs() {
        List<Session> tabs = sessionManager.getTabs();
        for (Session tab : tabs) {
            sessionManager.dropTab(tab.getId());
        }
    }

    @Override
    public void subscribe(final Observer observer) {
        if (onTabModelChangedListener == null) {
            onTabModelChangedListener = new OnTabModelChangedListener() {
                @Override
                void onTabModelChanged(Session tab) {
                    observer.onTabUpdate(tab);
                }

                @Override
                public void onTabCountChanged(int count) {
                    observer.onUpdate(sessionManager.getTabs());
                }
            };
        }
        sessionManager.addTabsViewListener(onTabModelChangedListener);
        sessionManager.addTabsChromeListener(onTabModelChangedListener);
    }

    @Override
    public void unsubscribe() {
        if (onTabModelChangedListener != null) {
            sessionManager.removeTabsViewListener(onTabModelChangedListener);
            sessionManager.removeTabsChromeListener(onTabModelChangedListener);
            onTabModelChangedListener = null;
        }
    }

    private static abstract class OnTabModelChangedListener implements TabsViewListener,
            TabsChromeListener {
        @Override
        public void onTabStarted(@NonNull Session tab) {
        }

        @Override
        public void onTabFinished(@NonNull Session tab, boolean isSecure) {
        }

        @Override
        public void onURLChanged(@NonNull Session tab, String url) {
            onTabModelChanged(tab);
        }

        @Override
        public boolean handleExternalUrl(String url) {
            return false;
        }

        @Override
        public void updateFailingUrl(@NonNull Session tab, String url, boolean updateFromError) {
            onTabModelChanged(tab);
        }

        @Override
        public void onProgressChanged(@NonNull Session tab, int progress) {
        }

        @Override
        public void onReceivedTitle(@NonNull Session tab, String title) {
            onTabModelChanged(tab);
        }

        @Override
        public void onReceivedIcon(@NonNull Session tab, Bitmap icon) {
            onTabModelChanged(tab);
        }

        @Override
        public void onFocusChanged(@Nullable Session tab, @Factor int factor) {

        }

        @Override
        public void onTabAdded(@NonNull Session tab, @Nullable Bundle arguments) {
        }

        @Override
        public void onTabCountChanged(int count) {
        }

        @Override
        public void onLongPress(@NonNull Session tab, TabView.HitTarget hitTarget) {
        }

        @Override
        public void onEnterFullScreen(@NonNull Session tab, @NonNull TabView.FullscreenCallback callback, @Nullable View fullscreenContent) {
        }

        @Override
        public void onExitFullScreen(@NonNull Session tab) {
        }

        @Override
        public boolean onShowFileChooser(@NonNull Session tab, TabView tabView, ValueCallback<Uri[]> filePathCallback, WebChromeClient.FileChooserParams fileChooserParams) {
            return false;
        }

        @Override
        public void onGeolocationPermissionsShowPrompt(@NonNull Session tab, String origin, GeolocationPermissions.Callback callback) {

        }

        abstract void onTabModelChanged(Session tab);
    }
}

File: app/src/main/java/org/mozilla/focus/tabs/tabtray/TabsSessionModel.kt
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 4; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.focus.tabs.tabtray

import android.graphics.Bitmap
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.webkit.GeolocationPermissions
import android.webkit.ValueCallback
import android.webkit.WebChromeClient

import org.mozilla.focus.BuildConfig
import org.mozilla.rocket.tabs.Session
import org.mozilla.rocket.tabs.SessionManager
import org.mozilla.rocket.tabs.SessionManager.Observer
import org.mozilla.rocket.tabs.TabView

import java.util.ArrayList

internal class TabsSessionModel(private val sessionManager: SessionManager) : TabTrayContract.Model {
    private var modelObserver: SessionModelObserver? = null

    private val tabs = ArrayList<Session>()

    override fun loadTabs(listener: TabTrayContract.Model.OnLoadCompleteListener?) {
        tabs.clear()
        tabs.addAll(sessionManager.getTabs())

        listener?.onLoadComplete()
    }

    override fun getTabs(): List<Session> {
        return tabs
    }

    override fun getFocusedTab(): Session? {
        return sessionManager.focusSession
    }

    override fun switchTab(tabPosition: Int) {
        if (tabPosition >= 0 && tabPosition < tabs.size) {
            val target = tabs[tabPosition]
            val latestTabs = sessionManager.getTabs()
            val exist = latestTabs.indexOf(target) != -1
            if (exist) {
                sessionManager.switchToTab(target.id)
            }
        } else {
            if (BuildConfig.DEBUG) {
                throw ArrayIndexOutOfBoundsException("index: " + tabPosition + ", size: " + tabs.size)
            }
        }
    }

    override fun removeTab(tabPosition: Int) {
        if (tabPosition >= 0 && tabPosition < tabs.size) {
            sessionManager.dropTab(tabs[tabPosition].id)
        } else {
            if (BuildConfig.DEBUG) {
                throw ArrayIndexOutOfBoundsException("index: " + tabPosition + ", size: " + tabs.size)
            }
        }
    }

    override fun clearTabs() {
        val tabs = sessionManager.getTabs()
        for (tab in tabs) {
            sessionManager.dropTab(tab.id)
        }
    }

    override fun subscribe(observer: TabTrayContract.Model.Observer) {
        if (this.modelObserver == null) {
            this.modelObserver = object : SessionModelObserver() {
                override fun onTabModelChanged(session: Session) {
                    observer.onTabUpdate(session)
                }

                override fun onSessionCountChanged(count: Int) {
                    observer.onUpdate(sessionManager.getTabs())
                }
            }
        }
        sessionManager.register(this.modelObserver as Observer)
    }

    override fun unsubscribe() {
        if (modelObserver != null) {
            sessionManager.unregister(modelObserver as Observer)
            modelObserver = null
        }
    }

    private abstract class SessionModelObserver : SessionManager.Observer, Session.Observer {
        var session: Session? = null

        override fun onUrlChanged(session: Session, url: String?) {
            session.let { onTabModelChanged(it) }
        }

        override fun updateFailingUrl(url: String?, updateFromError: Boolean) {
            session?.let { onTabModelChanged(it) }
        }

        override fun onTitleChanged(session: Session, title: String?) {
            session.let { onTabModelChanged(it) }
        }

        override fun onReceivedIcon(icon: Bitmap?) {
            session?.let { onTabModelChanged(it) }
        }

        override fun onFocusChanged(session: Session?, factor: SessionManager.Factor) {
            this.session?.let { it.unregister(this) }
            this.session = session
            this.session?.let { it.register(this) }
        }

        internal abstract fun onTabModelChanged(session: Session)
    }
}
